test/new-e2e/pkg/test/suite.go

```diff
@@ -0,0 +1,273 @@
+	envValue := reflect.ValueOf(env)
+
+	for _, field := range envFields {
+		if !field.IsExported() {
+			continue
+		}
+
+		importKeyFromTag := field.Tag.Get(importKey)
+		isImportable := field.Type.Implements(reflect.TypeOf((*components.Importable)(nil)).Elem())
+		isPtrImportable := reflect.PtrTo(field.Type).Implements(reflect.TypeOf((*components.Importable)(nil)).Elem())
+
+		// Produce meaningful error in case we have an importKey but field is not importable
+		if importKeyFromTag != "" && !isImportable {
+			return fmt.Errorf("resource named %s has %s key but does not implement Importable interface", field.Name, importKey)
+		}
+
+		if !isImportable && isPtrImportable {
+			return fmt.Errorf("resource named %s of type %T implements Importable on pointer receiver but is not a pointer", field.Name, field.Type)
+		}
+
+		if !isImportable {
+			continue
+		}
+
+		// Create zero-value if not created (pointer to struct)
+		fieldValue := envValue.Elem().FieldByIndex(field.Index)
+		if fieldValue.IsNil() {
+			fieldValue.Set(reflect.New(fieldValue.Type().Elem()))
+		}
+
+		// If resources == nil, we stop at creating zero value to be used in provisioner
+		if resources != nil {
+			importable := fieldValue.Interface().(components.Importable)
+			resourceKey := importable.Key()
+			if importKeyFromTag != "" {
+				resourceKey = importKeyFromTag
+			}
+			if resourceKey == "" {
+				return fmt.Errorf("resource named %s has no import key set and no annotation", field.Name)
+			}
+
+			if rawResource, found := resources[resourceKey]; found {
+				err := importable.Import(rawResource, fieldValue.Interface())
+				if err != nil {
+					return fmt.Errorf("failed to import resource named: %s with key: %s, err: %w", field.Name, resourceKey, err)
+				}
+			} else {
+				return fmt.Errorf("expected resource named: %s with key: %s but not returned by provisioners", field.Name, resourceKey)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (bs *BaseSuite[Env]) providerContext(opTimeout time.Duration) (context.Context, context.CancelFunc) {
+	var ctx context.Context
+	var cancel func()
+
+	if deadline, ok := bs.T().Deadline(); ok {
+		deadline = deadline.Add(-provisionerGracePeriod)
+		ctx, cancel = context.WithDeadlineCause(context.Background(), deadline, errors.New("go test timeout almost reached, cancelling provisioners"))
+	} else {
+		ctx, cancel = context.WithTimeout(context.Background(), opTimeout)
+	}
+
+	return ctx, cancel
+}
+
+//
+// Overriden methods
+//
+
+// BeforeTest is executed right before the test starts and receives the suite and test names as input.
+// This function is called by [testify Suite].
+//
+// If you override BeforeTest in your custom test suite type, the function must call [test.BaseSuite.BeforeTest].
+//
+// [testify Suite]: https://pkg.go.dev/github.com/stretchr/testify/suite
+func (bs *BaseSuite[Env]) BeforeTest(string, string) {
+	// Reset provisioners to original provisioners
+	bs.targetProvisioners = bs.params.provisioners
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement. -->
**🥜 nitpick**
Same here, for non-lazy env we should reconcile right away